# Executive Report - Kyle Lombardo

## Part I: Research Question

With the recent uptick in hospital readmissions due to the addition of COPD and Hip & Knee measurement and historical trends, there is a need to understand why there has been an historically high number of readmissions in the hospital system. This executive report builds off the previous reports and will attempt to answer the queston of whether there are certain demographics, locales, pre-existing conditions, or other variables that will result in a higher likelihood of readmission.

In the Exploratory Data Analysis report showed that there was a statistical relationship between 'Population', 'Children', 'Initial_days', and 'TotalCharge' with 'ReAdmis' when compared using a t-test. These results were similar to the first report which found that there was a relationship between 'ReAdmis', 'TotalCharge', 'Initial_days', and 'VitD_levels' when analyzed using PCA.

In a previous report using logistic regression, there was a very strong logistic relationship between 'ReAdmis' with 'Item8' and 'Initial_days' with an accuracy of 98%, auc of roc of 0.998, and a confusion matrix of:
<table>
    <thead>
        <tr>
            <th colspan='2'>Logistic Regression</th>
        </tr>
    </thead>    
    <tr>
        <td></td>
        <td><strong>Predict ReAdmis</strong></td>
        <td><strong>Predict Not</strong></td>
    </tr>
    <tr>
        <td>Actual ReAdmis</td>
        <td>1241</td>
        <td>25</td>
    </tr>
    <tr>
        <td>Actual Not</td>
        <td>25</td>
        <td>709</td>
    </tr>    
</table>

In another report, K-nearest neighbor proved to be less effective than logistic regression even with hyperparameter tuning for `n_neighbors`. KNN gave an accuracy score of 95%, auc of roc of 0.952, and a confusion matrix of:
<table>
    <thead>
        <tr>
            <th colspan='2'>K-nearest neighbor</th>
        </tr>
    </thead>    
    <tr>
        <td></td>
        <td><strong>Predict ReAdmis</strong></td>
        <td><strong>Predict Not</strong></td>
    </tr>
    <tr>
        <td>Actual ReAdmis</td>
        <td>1232</td>
        <td>47</td>
    </tr>
    <tr>
        <td>Actual Not</td>
        <td>43</td>
        <td>678</td>
    </tr>    
</table>

This report will use Random Forest, a supervised machine learning |algorithm, to see if it is a more accurate method than logistic regression and KNN in predicting whether a patient will be readmitted into the hospital after being discharged.

The data used for this report includes 10,000 patient record's responses to the following criteria from 'D207 D208 D209 Medical Data Consideration and Dictionary.pdf'. All strings that could not be re-expressed were removed as they would be impossible to use for random forest. Redundant data such as 'Lat' and 'Lng' compared to 'Zip' or meaningless data such as 'Customer_id' were removed. Lastly, because random forest is only compatible with continuous values, many features will need to be re-expressed as well.

(Criteria marked with a * were removed for this report)


Criteria | Data Type | Example | Description
----- | ----- | ----- | -----
*Case Order | Integer | 1034 | Index Values
*Customer_id	|String |‘D550524’|	Unique identifier for patient
*Interaction |	String|	‘8cd49b13-f45a-4b47-a2bd-173ffa932c2f’|	Unique Identifier for transaction, procedure, and admission
*UID	|String|	‘3a83ddb66e2ae73798bdf1d705dc0932’	|Unique Identifier for transaction, procedure, and admission
*City	|String	|‘Braggs’	|Patient city address from billing
*State	|String|	‘AL’|	Patient state address from billing
*County	|String|	‘Morgan’|	Patient county address from billing
Zip	|Integer|	35621	|Patient zip address from billing
*Lat	|Float	|-86.5404	|Patient latitude address from billing
*Lng	|Float	|-81.1272	|Patient longitude address from billing
Population	|Integer|	281	|Patient city address population from billing
Area	|String	|‘Urban’|	Patient address from billing zoning type
*Timezone|	String|	‘America/New_York’|	Patient address from billing time zone locale
*Job	|String	|‘Actuary’	|Patient or primary insurance holder’s occupation 
Children|	Integer|	2|	Number of children in patient’s household
Age	|Integer|	53	|Patient’s age
Income|	Float|	88126.93|	Patient or primary insurance holder’s yearly income
Marital|	String|	‘Married’|	Patient’s marital status
Gender	|String|	‘Female’	|Patient’s self-identification of gender
ReAdmis	|String	|‘Yes’	|Whether patient has been readmitted within a month of last release
VitD_level|	Float|	47.81348|	Patient’s vitamin d level (ng/mL)
Doc_visits|	Integer|	5	|Number of visits by primary physician during initial hospitalization
Full_meals_eaten|	Integer	|1|	Number of full meals the patient ate while in hospital.
VitD_supp|	Integer|	2	|Number of times vitamin d supplement was ministered to patient
Soft_drink	|String|	‘Yes’|	Whether patient consumes >= 3 soft drinks in a day
Initial_admin	|String	|‘Observation Admission’	|Means by which patient was admitted to hospital initially
HighBlood	|String	|‘Yes’	|Whether patient has high blood pressure
Stroke	|String|	‘No’	|Whether patient has had a stroke
Complication_risk|	String	|‘High’	|Complication risk level as determined by primary physician
Overweight|	String	|‘Yes’	|Whether patient is considered obese considering his or her demographics
Arthritis|	String	|‘No’|	Whether patient has arthritis
Diabetes	|String	|‘Yes’|	Whether patient has diabetes
Hyperlipidemia	|String|	‘No’|	Whether patient has hyperlipidemia
BackPain	|String	|‘Yes’|	Whether patient has chronic back pain
Anxiety	|String	|‘Yes’	|Whether patient has anxiety
Allergic_rhinitis	|String	|‘No’|	Whether patient has allergic rhinitis
Reflux_esophagitis	|String	|‘Yes’|	Whether patient has reflux esophagitis
Asthma	|String|	‘No’|	Whether patient has asthma
Services	|String|	‘CT Scan’	|Primary service patient received from hospital
Initial_days|	Float	|7.302395	|Number of days patient stayed in hospital on initial visit
TotalCharge	|Float	|2631.702|	Average amount charged to patient daily 
Additional_charges|	Float|	14382.23|	Average amount for miscellaneous procedures, treatments, medicines, etc

Patient’s opinion survey on rate of importance (1 = most important, 8 = least important)

Criteria | Data Type | Example | Description
----- | ----- | ----- | -----
Item1 |Integer| 3| Timely admission
Item2	|Integer|2|	Timely treatment
Item3	|Integer|5|	Timely visits
Item4	|Integer|6|	Reliability
Item5	|Integer|1|	Options
Item6	|Integer|2|	Hours of treatment
Item7	|Integer|6|	Courteous staff
Item8	|Integer|4|	Evidence of active listening from doctor

## Part II: Method Justification

Random forest is a useful machine learning tool used to predict either categorical and regression target variables using multiple continuous variables in a data set. Random forest is made of a user defined number of decision trees each with different random entries of the same size. It accomplishes this by randomly selecting a fixed number of entries, in this case 10,000, with replacement meaning that the same entry can be used multiple times. In addition, random forest also picks features from a random subset for each decision tree. Both of these attributes of random forest ensures that the method for the decision making is different for each unique decision tree.<sup>1</sup>

When it comes time for a prediction, random forest will send the entry in question through each decision tree, tally up all the votes from each decision tree, and the highest vote will be the final prediction. In this particular case, it will predict those patients that are or are not readmitted.

Some assumptions for random forest to work properly is that all data fed into the model as predictors must be cleaned with no missing data. All feature variables must be continuous data so no categorical data in the form of strings can be used in the data set. All categorical data will be converted using one-hot encoding through the python method `pd.get_dummies()`. This will result in a much higher number of features for the final data set as each categorical variable will be split into n number of features where n = number of unique categories - 1.<sup>2</sup>

In this report Python will be used. While R could be used, the python scikit learn library has many useful tools including hyperparameter tuning with gridsearch, pipelines, and modeling techniques that makes random forest easy and quick to use. 

As for libraries used in the report, the following will be utilized. Pandas and numpy are crucial tools for importing, re-expressing, and manipulating the data. Scikit Learn is used for multiple things in this report including:

<ol>
    <li><strong><em>Pipeline</em></strong>: a way to group many scikit learn methods together in a clean and quick fashion. </li>
    <li><strong><em>GridSearchCV</em></strong>: used for hyperparameter tuning. It will try a number of different combinations of hyperparameters in a given range and output the best one according to a score. This score will be set to accuracy for random forest.</li>
    <li><strong><em>train_test_split</em></strong>: used in order to split up the feature and target variables into separate training and testing set. This is best practice in order to test the model using outside information, or information not used in the training set.</li>
    <li><strong><em>RandomForestClassifier</em></strong>: random forest model method already described above.</li>
    <li><strong><em>classification_report</em></strong>: gives a summary of accuracy, recall, and precision among other parameters.</li>
    <li><strong><em>confusion_matrix</em></strong>: a tabled statistical summary of true positives, true negatives, false positives, and false negatives of the prediction values vs the actual values</li>
    <li><strong><em>roc_auc_score</em></strong>: area under the curve of the graph plotting the true positive rate vs. the false positive rate also known as receiver operating curve. A perfect score includes only true positive values and gives an area of 1.0</li>
</ol>

The data set 'df' is imported here using pandas.  

In [106]:
import pandas as pd
import numpy as np

from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score

raw_data = pd.read_csv('medical_clean.csv')

## Part III: Data Preparation

As stated above, the data must be in the correct format and data types in order to be run properly in random forest. The data set will need to undergo a data cleaning step as follows: 

First, all variables not being used will be removed from the data set. Only numeric variables or variables that can be re-expressed as numeric will remain. This is the only data type random forest is able to use. Next, re-expressions will be performed using one-hot encoding using the `get_dummies()`, a pandas method, on all categorical variables including the target variable. This will break up all categorical variables into separate binary variables. The first binary variable is dropped from each previous variable due to being redundant. 

Second, any missing values must be imputed using either an average or the most frequent value.

### Removing unnecessary and redundant data

In [107]:
df = raw_data.drop(['CaseOrder', 'Customer_id', 'Interaction', 'UID', 'City', 'State', \
                    'County', 'Lat', 'Lng', 'TimeZone','Job'], axis=1)
df = pd.get_dummies(df, drop_first=True)

### Re-expressing data to numeric data types and missing values

For a quick way to see how many missing values are in the data set `df.info()` can be used. It looks as if there is no missing data values. Imputation will not need to be performed on this data set. Also, all data types are either integers or floats and are ready to be run using random forest.

With the data set `df` cleaned and in correct type format, the mean and variance for each feature can be calculated using `df.describe()`.

In [108]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 48 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Zip                                  10000 non-null  int64  
 1   Population                           10000 non-null  int64  
 2   Children                             10000 non-null  int64  
 3   Age                                  10000 non-null  int64  
 4   Income                               10000 non-null  float64
 5   VitD_levels                          10000 non-null  float64
 6   Doc_visits                           10000 non-null  int64  
 7   Full_meals_eaten                     10000 non-null  int64  
 8   vitD_supp                            10000 non-null  int64  
 9   Initial_days                         10000 non-null  float64
 10  TotalCharge                          10000 non-null  float64
 11  Additional_charges           

In [109]:
df.describe(include='all')

Zip     Population      Children           Age         Income  \
count  10000.000000   10000.000000  10000.000000  10000.000000   10000.000000   
mean   50159.323900    9965.253800      2.097200     53.511700   40490.495160   
std    27469.588208   14824.758614      2.163659     20.638538   28521.153293   
min      610.000000       0.000000      0.000000     18.000000     154.080000   
25%    27592.000000     694.750000      0.000000     36.000000   19598.775000   
50%    50207.000000    2769.000000      1.000000     53.000000   33768.420000   
75%    72411.750000   13945.000000      3.000000     71.000000   54296.402500   
max    99929.000000  122814.000000     10.000000     89.000000  207249.100000   

        VitD_levels    Doc_visits  Full_meals_eaten     vitD_supp  \
count  10000.000000  10000.000000      10000.000000  10000.000000   
mean      17.964262      5.012200          1.001400      0.398900   
std        2.017231      1.045734          1.008117      0.628505   
min        9.806483      1.000000          0.000000      0.000000   
25%       16.626439      4.000000          0.000000      0.000000   
50%       17.951122      5.000000          1.000000      0.000000   
75%       19.347963      6.000000          2.000000      1.000000   
max       26.394449      9.000000          7.000000      5.000000   

       Initial_days  ...  Diabetes_Yes  Hyperlipidemia_Yes  BackPain_Yes  \
count  10000.000000  ...   10000.00000        10000.000000  10000.000000   
mean      34.455299  ...       0.27380            0.337200      0.411400   
std       26.309341  ...       0.44593            0.472777      0.492112   
min        1.001981  ...       0.00000            0.000000      0.000000   
25%        7.896215  ...       0.00000            0.000000      0.000000   
50%       35.836244  ...       0.00000            0.000000      0.000000   
75%       61.161020  ...       1.00000            1.000000      1.000000   
max       71.981490  ...       1.00000            1.000000      1.000000   

        Anxiety_Yes  Allergic_rhinitis_Yes  Reflux_esophagitis_Yes  \
count  10000.000000           10000.000000            10000.000000   
mean       0.321500               0.394100                0.413500   
std        0.467076               0.488681                0.492486   
min        0.000000               0.000000                0.000000   
25%        0.000000               0.000000                0.000000   
50%        0.000000               0.000000                0.000000   
75%        1.000000               1.000000                1.000000   
max        1.000000               1.000000                1.000000   

        Asthma_Yes  Services_CT Scan  Services_Intravenous  Services_MRI  
count  10000.00000      10000.000000          10000.000000  10000.000000  
mean       0.28930          0.122500              0.313000      0.038000  
std        0.45346          0.327879              0.463738      0.191206  
min        0.00000          0.000000              0.000000      0.000000  
25%        0.00000          0.000000              0.000000      0.000000  
50%        0.00000          0.000000              0.000000      0.000000  
75%        1.00000          0.000000              1.000000      0.000000  
max        1.00000          1.000000              1.000000      1.000000  

[8 rows x 48 columns]

Finally, for the last step in the data preparation stage, the cleaned and prepared dataset is exported for the final report.

In [110]:
df.to_csv(r'C:\Users\kylel\OneDrive\Desktop\Learning\School\D209 - Data Mining I\PA2\prepped_data.csv')

## Part IV: Analysis

With the data cleaned and prepared, it is time to run the random forest model. The data will need to be split into `X` for features and `y` as the target. 

First, a pipeline will be set up to make things a bit more organized. This pipeline will use `RandomForestClassifier()` and allow the use of `GridSearchCV` to be run.  

Next, the data will be split into training and testing sets using `train_test_split()` with 80% going to the training set and the remaining 20% to the testing set. This is a crucial step in the modeling process. If all of `X` is used, there would be no way to test the model's accuracy. Splitting the data into training and testing sets allows for the ability to use unseen data (the test set) to be used to test for accuracy.

The X_train and y_train (along with the X_test and y_test) are merged together using the pandas merge function and exported to separate csvs for testing and training. 

After, a `GridSearchCV()` will be used to tune the `n_estimators` hyperparameter. This will cycle through a range of a number of decision trees for the `n_estimators` values that gives the highest accuracy. This was already done a few times ahead of time to find the best range to iterate through. GridSearch typically has a default scoring of R<sup>2</sup>, but the best scoring parameter for a binary target would be accuracy.

Finally, the model will be fit and ready to analyze and be examined.<sup>2</sup>

In [111]:
X = df.drop(['ReAdmis_Yes'], axis=1)
y = df['ReAdmis_Yes']

In [112]:
steps = [('RFC', RandomForestClassifier())]
pipeline = Pipeline(steps)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=41)

In [113]:
training_set = pd.merge(X_train, pd.DataFrame(y_train), left_index=True, right_index=True)
training_set.to_csv(r'C:\Users\kylel\OneDrive\Desktop\Learning\School\D209 - Data Mining I\PA2\training_data.csv')

testing_set = pd.merge(X_test, pd.DataFrame(y_test), left_index=True, right_index=True)
testing_set.to_csv(r'C:\Users\kylel\OneDrive\Desktop\Learning\School\D209 - Data Mining I\PA2\testing_data.csv')

In [119]:
parameters = {'RFC__n_estimators': np.arange(210,230,2)}
cv = GridSearchCV(pipeline, parameters, cv=3, scoring='accuracy')

cv.fit(X_train, y_train)
y_pred = cv.predict(X_test)

In [120]:
print(classification_report(y_test, y_pred))
print('Tuned Model Parameters: {}'.format(cv.best_params_))

              precision    recall  f1-score   support

           0       0.98      0.99      0.98      1279
           1       0.98      0.97      0.97       721

    accuracy                           0.98      2000
   macro avg       0.98      0.98      0.98      2000
weighted avg       0.98      0.98      0.98      2000

Tuned Model Parameters: {'RFC__n_estimators': 224}


In [121]:
confuse_matrix = confusion_matrix(y_test, y_pred)
confuse_matrix

array([[1262,   17],
       [  23,  698]], dtype=int64)

In [122]:
roc = roc_auc_score(y_test, y_pred)
print('Auc score: {:.3f}'.format(roc))

Auc score: 0.977


In [123]:
from sklearn.metrics import mean_squared_error as MSE
mse = MSE(y_test, y_pred)
print(mse)

0.02


## Part V: Data Summary and Implications

The range in the `parameters` variable has been tuned a number of times to find the value that works best for this model. With the number of estimators set to 212 (although this will differ from run to run due to the randomness of random forest) providing the highest accuracy for the test set.

A classification_report was also used to see what sort of accuracy this particular model gives. 98% is even with the logistic regression when using all features whether they help or hinder the prediction. When comparing confusion matrices, the random forest does perform slighty better with lower values for both false positives and false negatives. Logistic regression does have a higher `roc_auc_score` with 0.998 vs random forest's 0.977 (which can also differ slightly). This model also gives a mean squared error of 0.019 (yet again, can differ slightly) but mse is not an appropriate metric for classification and should be ignored. 

However, with all that has been said, random foreset is able to make a solid, quick prediction (in the sense of programming and hyper parameter tuning) of whether a patient may or may not be readmitted to the hospital. One of the biggest limitations of random forest is the speed at which the algorithms works. Creating such a high number of decision trees takes a long toll on the speed of the modeling. Another limitation is the data preparation that needs to happen before it can even be run including removing an null values and its ability to only use numeric values.<sup>3</sup>

The recommended actions at this point are to either use the logitistic regression covered in a previous report or this random forest model given that they both have similar predictive outputs. The logistic model took longer to tune due to feature reduction but can perform quickly now. The random forest does take a bit longer to perform but gives similar results.

## Resources

1. Elie Kawerk, <em>Machine Learning with Tree-Based Models in Python</em>, DataCamp, accessed 22 October 2021, <<https://campus.datacamp.com/courses/machine-learning-with-tree-based-models-in-python/bagging-and-random-forests?ex=7>>.
2. Hugo Bowne-Anderson, <em>Supervised Learning with scikit-learn</em>, DataCamp, accessed 22 October 2021, <<https://app.datacamp.com/learn/courses/supervised-learning-with-scikit-learn>>.
3. Niklas Donges, <em>A Complete Guide to the Random Forest Algorithm</em>, Built In Beta, accessed 22, October 2021, <<https://builtin.com/data-science/random-forest-algorithm#procon>>.
